lib/sqlalchemy/dialects/sqlite/base.py

```diff
@@ -402,6 +402,253 @@ def set_sqlite_pragma(dbapi_connection, connection_record):
+        unique=True,
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        index_elements=['id'],
+        set_=dict(data='updated value')
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        index_elements=[my_table.c.id],
+        set_=dict(data='updated value')
+    )
+
+* When using :paramref:`.Insert.on_conflict_do_update.index_elements` to
+  infer an index, a partial index can be inferred by also specifying the
+  use the :paramref:`.Insert.on_conflict_do_update.index_where` parameter::
+
+    from sqlalchemy.dialects.sqlite import insert
+
+    unique_partial_index = schema.Index(
+        "unique_partial_index_name",
+        my_table.c.user_email,
+        unique=True,
+        sqlite_where=my_table.c.user_email.like('@gmail.com'),
+    )
+
+    stmt = insert(my_table).values(user_email='a@b.com', data='inserted data')
+    stmt = stmt.on_conflict_do_update(
+        index_elements=[my_table.c.user_email],
+        index_where=my_table.c.user_email.like('%@gmail.com'),
+        set_=dict(data=stmt.excluded.data)
+        )
+    conn.execute(stmt)
+
+* The :paramref:`.Insert.on_conflict_do_update.constraint` argument is
+  used to specify a constraint directly rather than inferring it.  This can
+  either be the name of the underlying columns of a UNIQUE constraint or
+  a PRIMARY KEY constraint. SQLite does not support specifying the names of
+  an UNIQUE or PRIMARY KEY constraint and INDEX within the ON CONFLICT Clause.
+
+  That is, the following will not work::
+    unique_constraint_pk = schema.PrimaryKeyConstraint(
+        my_table.c.id, name="uq_user_pk"
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        constraint='uq_user_pk',
+        set_=dict(data='updated value')
+    )
+
+ But the statements below will work::
+    do_update_stmt_A = insert_stmt.on_conflict_do_update(
+        constraint=unique_constraint_pk,
+        set_=dict(id=1, data='updated value')
+    )
+
+    do_update_stmt_B = insert_stmt.on_conflict_do_update(
+        constraint='id',
+        set_=dict(id=1, data='updated value')
+    )
+
+* The :paramref:`.Insert.on_conflict_do_update.constraint` argument may
+  also refer to a SQLAlchemy construct representing a constraint,
+  e.g. :class:`.UniqueConstraint` or :class:`.PrimaryKeyConstraint`.
+  In this use, only the name of the constraint's first column is used directly.
+  If the constraint is unnamed, then inference will be used,
+  where the expressions and optional WHERE clause of the constraint will
+  be spelled out in the construct::
+
+    unique_constraint = schema.UniqueConstraint(
+        my_table.c.user_email, name="uq_user_email"
+    )
+
+    do_update_stmt = insert_stmt.on_conflict_do_update(
+        constraint=unique_constraint,
+        set_=dict(data='updated value')
+    )
+
+``ON CONFLICT...DO UPDATE`` is used to perform an update of the already
+existing row, using any combination of new values as well as values
```

Comment:
**mike bayer** (zzzeek) wrote:

then put a little header here:

The Set Clause
^^^^^^^^^^^^^^

just because these docs are long

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2230